File: sample/src/main/java/com/urbanairship/sample/preference/SettingsFragment.java
/* Copyright Airship and Contributors */

package com.urbanairship.sample.preference;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.urbanairship.debug.DebugActivity;
import com.urbanairship.sample.R;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import androidx.core.view.MenuProvider;
import androidx.fragment.app.Fragment;
import androidx.navigation.Navigation;
import androidx.navigation.ui.NavigationUI;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

import static com.urbanairship.debug.R.*;

/**
 * Settings fragment.
 *
 * Wraps the PreferenceFragment.
 */
public class SettingsFragment extends Fragment {

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_settings, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (savedInstanceState == null) {
            getChildFragmentManager().beginTransaction()
                                     .replace(R.id.preference_placeholder, new PreferenceFragment())
                                     .commitNow();
        }

        Toolbar toolbar = view.findViewById(R.id.toolbar);
        NavigationUI.setupWithNavController(toolbar, Navigation.findNavController(view));
    }

    /**
     * PreferenceFragmentCompat
     */
    public static class PreferenceFragment extends PreferenceFragmentCompat {

        private static final String TAGS_KEY = "tags";
        private static final String DEBUG_KEY = "debug";

        @Override
        public void onCreatePreferences(Bundle bundle, String s) {
            addPreferencesFromResource(R.xml.preferences);
        }

        @Override
        public boolean onPreferenceTreeClick(@NonNull Preference preference) {
            View view = requireView();

            switch (preference.getKey()) {
                case TAGS_KEY:
                    Navigation.findNavController(view).navigate(R.id.tagsFragment);
                    return true;
                case DEBUG_KEY:
                    requireActivity().startActivity(new Intent(requireContext(), DebugActivity.class)
                            .putExtra("includeBackButton", true));
                    return true;
            }

            return super.onPreferenceTreeClick(preference);
        }

    }

}

File: a11y/src/main/java/com/urbanairship/devapp/preference/SettingsFragment.kt
/* Copyright Airship and Contributors */
package com.urbanairship.devapp.preference

import android.os.Bundle
import android.view.View
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.fragment.app.commitNow
import androidx.navigation.fragment.findNavController
import androidx.navigation.ui.setupWithNavController
import com.urbanairship.preferencecenter.ui.PreferenceCenterFragment
import com.urbanairship.devapp.MainActivity.APP_BAR_CONFIGURATION
import com.urbanairship.devapp.R

class SettingsFragment : Fragment(R.layout.fragment_settings) {

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val toolbar: Toolbar = view.findViewById(R.id.toolbar)
        toolbar.setupWithNavController(findNavController(), APP_BAR_CONFIGURATION)

        val prefCenterId = arguments?.getString("pref_center_id") ?: "app_default"

        if (savedInstanceState == null) {
            val fragment = PreferenceCenterFragment.create(prefCenterId).apply {
                setOnDisplayPreferenceCenterListener(OnDisplayListener(this, toolbar))
            }

            childFragmentManager.commitNow {
                setReorderingAllowed(true)
                replace(R.id.fragment_container_view, fragment)
            }
        } else {
            val fragment = (childFragmentManager.findFragmentById(R.id.fragment_container_view) as PreferenceCenterFragment)
            // Reset our listener on the restored fragment.
            fragment.setOnDisplayPreferenceCenterListener(OnDisplayListener(fragment, toolbar))
        }
    }

    private class OnDisplayListener(
        private val fragment: PreferenceCenterFragment,
        private val toolbar: Toolbar
    ) : PreferenceCenterFragment.OnDisplayPreferenceCenterListener {
        override fun onDisplayPreferenceCenter(title: String?, description: String?): Boolean {
            toolbar.title = title
            if (description != null) {
                fragment.showHeaderItem(title = null, description = description)
            }
            return true
        }
    }
}
